lib/em-hiredis/connection_manager.rb

```diff
@@ -0,0 +1,129 @@
+require 'uri'
+
+module EventMachine::Hiredis
+  # Emits the following events
+  #
+  # * :connected - on successful connection or reconnection
+  # * :reconnected - on successful reconnection
+  # * :disconnected - no longer connected, when previously in connected state
+  # * :reconnect_failed(failure_number) - a reconnect attempt failed
+  #     This event is passed number of failures so far (1,2,3...)
+  class ConnectionManager
+    include EventEmitter
+
+    TRANSITIONS = [
+      # first connect call
+      [ :initial, :connecting ],
+      # TCP connect, or initialisation commands fail
+      [ :connecting, :disconnected ],
+      # Connection ready for use by clients
+      [ :connecting, :connected ],
+      # connection lost
+      [ :connected, :disconnected ],
+      # attempting automatic reconnect
+      [ :disconnected, :connecting ],
+      # all automatic reconnection attempts failed
+      [ :disconnected, :failed ],
+      # manual call of reconnect after failure
+      [ :failed, :connecting ],
+    ]
+
+    def initialize(em, connection_factory)
+      @em = em
+      @connection_factory = connection_factory
+
+      @reconnect_attempt = 0
+
+      @sm = StateMachine.new
+      TRANSITIONS.each { |t| @sm.transition(*t) }
+
+      @sm.on(:connecting, &method(:connect_internal))
+      @sm.on(:connected, &method(:connected))
+      @sm.on(:disconnected, &method(:disconnected))
+      @sm.on(:failed, &method(:perm_failure))
```

Comment:
I know this is a small thing, but could you rename these to: on_connecting, on_connected...